core/CronArchive.php

```diff
@@ -938,6 +934,77 @@ public function isThereExistingValidPeriod(Parameters $params)
         return !empty($idArchive);
     }
 
+    // public for tests
+    public function repairInvalidationsIfNeeded($archiveToProcess)
+    {
+        $table = Common::prefixTable('archive_invalidations');
+
+        $bind = [
+            $archiveToProcess['idsite'],
+            $archiveToProcess['name'],
+            $archiveToProcess['period'],
+            $archiveToProcess['date1'],
+            $archiveToProcess['date2'],
+        ];
+
+        $reportClause = '';
+        if (!empty($archiveToProcess['report'])) {
+            $reportClause = " AND report = ?";
+            $bind[] = $archiveToProcess['report'];
+        }
+
+        $sql = "SELECT DISTINCT period FROM `$table` WHERE idsite = ? AND name = ? AND period > ? AND ? >= date1 AND date2 >= ? $reportClause";
+
+        $higherPeriods = Db::fetchAll($sql, $bind);
+        $higherPeriods = array_column($higherPeriods, 'period');
+        $higherPeriods = array_flip($higherPeriods);
+
+        $invalidationsToInsert = [];
+        foreach (Piwik::$idPeriods as $label => $id) {
+            // lower period than the one we're processing or range, don't care
+            if ($id <= $archiveToProcess['period'] || $label == 'range') {
+                continue;
+            }
+
+            if (isset($higherPeriods[$id])) { // period exists in table
+                continue;
+            }
+
+            // archive is for week that is over two months, we don't need to care about the month
+            if ($label == 'month'
+                && Date::factory($archiveToProcess['date1'])->toString('m') != Date::factory($archiveToProcess['date2'])->toString('m')
+            ) {
+                continue;
+            }
+
+            $period = Period\Factory::build($label, $archiveToProcess['date1']);
+
+            $invalidationToInsert = [
+                'idarchive' => null,
+                'name' => $archiveToProcess['name'],
+                'report' => $archiveToProcess['report'],
+                'idsite' => $archiveToProcess['idsite'],
+                'date1' => $period->getDateStart()->getDatetime(),
+                'date2' => $period->getDateEnd()->getDatetime(),
+                'period' => $id,
+                'ts_invalidated' => $archiveToProcess['ts_invalidated'],
```

Comment:
should this have maybe the current date? Just to be more clear when it was inserted. Or is it so it's included to be archived in this archive run maybe? (as we maybe do somewhere `and ts_invalidated < idSiteStartTime`)

Reply:
Yes, it's done that way so we process it in this batch. (getNextInvalidatedArchive does the `ts_invalidated < idSiteStartTime`)